pkg/reconciler/apibinding/apibinding_controller.go

```diff
@@ -0,0 +1,348 @@
+		AddFunc:    func(obj interface{}) { c.enqueueAPIResourceSchema(obj) },
+		UpdateFunc: func(_, obj interface{}) { c.enqueueAPIResourceSchema(obj) },
+		DeleteFunc: func(obj interface{}) { c.enqueueAPIResourceSchema(obj) },
+	})
+
+	//if err := crdInformer.Informer().GetIndexer().AddIndexers(cache.Indexers{
+	//	indexAPIResourceSchemasByCRD: func(obj interface{}) ([]string, error) {
+	//		crd, ok := obj.(*apiextensionsv1.CustomResourceDefinition)
+	//		if !ok {
+	//			return []string{}, fmt.Errorf("obj is supposed to be a CRD, but is %T", obj)
+	//		}
+	//		if crd.Annotations[annotationSchemaClusterKey] != "" && crd.Annotations[annotationSchemaNameKey] != "" {
+	//			return []string{clusters.ToClusterAwareKey(crd.Annotations[annotationSchemaClusterKey], crd.Annotations[annotationSchemaNameKey])}, nil
+	//		}
+	//		return []string{}, nil
+	//	},
+	//}); err != nil {
+	//	return nil, fmt.Errorf("error add CRD indexes: %w", err)
+	//}
+
+	apiExportInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueueAPIExport(obj) },
+		UpdateFunc: func(_, obj interface{}) { c.enqueueAPIExport(obj) },
+		DeleteFunc: func(obj interface{}) { c.enqueueAPIExport(obj) },
+	})
+
+	if err := c.apiExportsIndexer.AddIndexers(cache.Indexers{
+		indexAPIExportsByAPIResourceSchema: indexAPIExportByAPIResourceSchemas,
+	}); err != nil {
+		return nil, fmt.Errorf("error add CRD indexes: %w", err)
+	}
+
+	return c, nil
+}
+
+// controller reconciles APIBindings. It creates and maintains CRDs associated with an APIBinding's reference. It
+// also watches CRDs, APIResourceSchemas, and APIExports to ensure whenever objects related to an APIBinding are
+// updated, the APIBinding is rereconciled.
+type controller struct {
+	queue workqueue.RateLimitingInterface
+
+	crdClusterClient apiextensionclientset.ClusterInterface
+	kcpClusterClient kcpclient.ClusterInterface
+
+	apiBindingsLister        apislisters.APIBindingLister
+	apiBindingsIndexer       cache.Indexer
+	apiExportsLister         apislisters.APIExportLister
+	apiExportsIndexer        cache.Indexer
+	apiResourceSchemaLister  apislisters.APIResourceSchemaLister
+	apiResourceSchemaIndexer cache.Indexer
+	crdLister                apiextensionslisters.CustomResourceDefinitionLister
+	crdIndexer               cache.Indexer
+
+	deletedCRDTracker *lockedStringSet
+}
+
+// enqueueAPIBinding enqueues an APIBinding .
+func (c *controller) enqueueAPIBinding(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	klog.Infof("Queueing APIBinding %q", key)
+	c.queue.Add(key)
+}
+
+// enqueueAPIExport enqueues maps an APIExport to APIBindings for enqueuing.
+func (c *controller) enqueueAPIExport(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	klog.Infof("Mapping APIExport %q", key)
```

Comment:
(No action required) Does this happen so infrequently that Infof is suggested?

Reply:
infrequent I think